File: src/CalcManager/CEngine/CalcInput.rs
use std::cmp::Ordering;

const MAX_STRLEN: usize = 84;
const C_NUM_MAX_DIGITS: usize = MAX_STRLEN;
const C_EXP_MAX_DIGITS: usize = 4;

pub struct CalcNumSec {
    value: String,
    is_negative: bool,
}

impl CalcNumSec {
    pub fn new() -> Self {
        Self {
            value: String::new(),
            is_negative: false,
        }
    }

    pub fn clear(&mut self) {
        self.value.clear();
        self.is_negative = false;
    }

    pub fn is_empty(&self) -> bool {
        self.value.is_empty()
    }

    pub fn is_negative(&self) -> bool {
        self.is_negative
    }

    pub fn set_negative(&mut self, is_negative: bool) {
        self.is_negative = is_negative;
    }
}

pub struct CalcInput {
    base: CalcNumSec,
    exponent: CalcNumSec,
    has_exponent: bool,
    has_decimal: bool,
    dec_pt_index: usize,
    dec_symbol: char,
}

impl CalcInput {
    pub fn new(dec_symbol: char) -> Self {
        Self {
            base: CalcNumSec::new(),
            exponent: CalcNumSec::new(),
            has_exponent: false,
            has_decimal: false,
            dec_pt_index: 0,
            dec_symbol,
        }
    }

    pub fn clear(&mut self) {
        self.base.clear();
        self.exponent.clear();
        self.has_exponent = false;
        self.has_decimal = false;
        self.dec_pt_index = 0;
    }

    pub fn try_toggle_sign(&mut self, is_integer_mode: bool, max_num_str: &str) -> bool {
        if self.base.is_empty() {
            self.base.set_negative(false);
            self.exponent.set_negative(false);
        } else if self.has_exponent {
            self.exponent.set_negative(!self.exponent.is_negative());
        } else {
            if is_integer_mode && self.base.is_negative() {
                if self.base.value.len() >= max_num_str.len() && self.base.value.chars().last() > max_num_str.chars().last() {
                    return false;
                }
            }
            self.base.set_negative(!self.base.is_negative());
        }
        true
    }

    pub fn try_add_digit(&mut self, value: u32, radix: u32, is_integer_mode: bool, max_num_str: &str, word_bit_width: i32, max_digits: usize) -> bool {
        let ch_digit = if value < 10 {
            (b'0' + value as u8) as char
        } else {
            (b'A' + value as u8 - 10) as char
        };

        let (p_num_sec, max_count) = if self.has_exponent {
            (&mut self.exponent, C_EXP_MAX_DIGITS)
        } else {
            let mut max_count = max_digits;
            if self.has_decimal {
                max_count += 1;
            }
            if !self.base.is_empty() && self.base.value.starts_with('0') {
                max_count += 1;
            }
            (&mut self.base, max_count)
        };

        if p_num_sec.is_empty() && value == 0 {
            return true;
        }

        if p_num_sec.value.len() < max_count {
            p_num_sec.value.push(ch_digit);
            return true;
        }

        if is_integer_mode && p_num_sec.value.len() == max_count && !self.has_exponent {
            let allow_extra_digit = match radix {
                8 => match word_bit_width % 3 {
                    1 => p_num_sec.value.starts_with('1'),
                    2 => p_num_sec.value.starts_with(|c| c <= '3'),
                    _ => false,
                },
                10 => {
                    if p_num_sec.value.len() < max_num_str.len() {
                        match p_num_sec.value.cmp(&max_num_str[..p_num_sec.value.len()]) {
                            Ordering::Less => true,
                            Ordering::Equal => {
                                let last_char = max_num_str.chars().nth(p_num_sec.value.len()).unwrap();
                                ch_digit <= last_char || (p_num_sec.is_negative() && ch_digit <= last_char + 1)
                            }
                            Ordering::Greater => false,
                        }
                    } else {
                        false
                    }
                }
                _ => false,
            };

            if allow_extra_digit {
                p_num_sec.value.push(ch_digit);
                return true;
            }
        }

        false
    }

    pub fn try_add_decimal_pt(&mut self) -> bool {
        if self.has_decimal || self.has_exponent {
            return false;
        }

        if self.base.is_empty() {
            self.base.value.push('0');
        }

        self.dec_pt_index = self.base.value.len();
        self.base.value.push(self.dec_symbol);
        self.has_decimal = true;

        true
    }

    pub fn has_decimal_pt(&self) -> bool {
        self.has_decimal
    }

    pub fn try_begin_exponent(&mut self) -> bool {
        self.try_add_decimal_pt();

        if self.has_exponent {
            return false;
        }

        self.has_exponent = true;
        true
    }

    pub fn backspace(&mut self) {
        if self.has_exponent {
            if !self.exponent.is_empty() {
                self.exponent.value.pop();
                if self.exponent.is_empty() {
                    self.exponent.clear();
                }
            } else {
                self.has_exponent = false;
            }
        } else {
            if !self.base.is_empty() {
                self.base.value.pop();
                if self.base.value == "0" {
                    self.base.value.pop();
                }
            }

            if self.base.value.len() <= self.dec_pt_index {
                self.has_decimal = false;
                self.dec_pt_index = 0;
            }

            if self.base.is_empty() {
                self.base.clear();
            }
        }
    }

    pub fn set_decimal_symbol(&mut self, dec_symbol: char) {
        if self.dec_symbol != dec_symbol {
            self.dec_symbol = dec_symbol;

            if self.has_decimal {
                self.base.value.replace_range(self.dec_pt_index..=self.dec_pt_index, &dec_symbol.to_string());
            }
        }
    }

    pub fn is_empty(&self) -> bool {
        self.base.is_empty() && !self.has_exponent && self.exponent.is_empty() && !self.has_decimal
    }

    pub fn to_string(&self, radix: u32) -> String {
        if self.base.value.len() > MAX_STRLEN || (self.has_exponent && self.exponent.value.len() > MAX_STRLEN) {
            return String::new();
        }

        let mut result = String::new();

        if self.base.is_negative() {
            result.push('-');
        }

        if self.base.is_empty() {
            result.push('0');
        } else {
            result.push_str(&self.base.value);
        }

        if self.has_exponent {
            if !self.has_decimal {
                result.push(self.dec_symbol);
            }

            result.push(if radix == 10 { 'e' } else { '^' });
            result.push(if self.exponent.is_negative() { '-' } else { '+' });

            if self.exponent.is_empty() {
                result.push('0');
            } else {
                result.push_str(&self.exponent.value);
            }
        }

        if result.len() > C_NUM_MAX_DIGITS * 2 + 4 {
            return String::new();
        }

        result
    }

    pub fn to_rational(&self, radix: u32, precision: i32) -> Rational {
        let rat = string_to_rat(self.base.is_negative(), &self.base.value, self.exponent.is_negative(), &self.exponent.value, radix, precision);
        if rat.is_none() {
            return Rational::default();
        }

        let result = Rational::from_rat(rat.unwrap());
        destroy_rat(rat.unwrap());

        result
    }
}

File: src/CalcManager/CEngine/CalcInput.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include <sstream>
#include "Header Files/CalcEngine.h"

using namespace std;
using namespace CalcEngine;

constexpr int C_NUM_MAX_DIGITS = MAX_STRLEN;
constexpr int C_EXP_MAX_DIGITS = 4;

void CalcNumSec::Clear()
{
    value.clear();
    m_isNegative = false;
}

void CalcInput::Clear()
{
    m_base.Clear();
    m_exponent.Clear();
    m_hasExponent = false;
    m_hasDecimal = false;
    m_decPtIndex = 0;
}

bool CalcInput::TryToggleSign(bool isIntegerMode, wstring_view maxNumStr)
{
    // Zero is always positive
    if (m_base.IsEmpty())
    {
        m_base.IsNegative(false);
        m_exponent.IsNegative(false);
    }
    else if (m_hasExponent)
    {
        m_exponent.IsNegative(!m_exponent.IsNegative());
    }
    else
    {
        // When in integer only mode, it isn't always allowed to toggle, as toggling can cause the num to be out of
        // bounds. For eg. in byte -128 is valid, but when it toggled it becomes 128, which is more than 127.
        if (isIntegerMode && m_base.IsNegative())
        {
            // Decide if this additional digit will fit for the given bit width
            if (m_base.value.size() >= maxNumStr.size() && m_base.value.back() > maxNumStr.back())
            {
                // Last digit is more than the allowed positive number. Fail
                return false;
            }
        }
        m_base.IsNegative(!m_base.IsNegative());
    }

    return true;
}

bool CalcInput::TryAddDigit(unsigned int value, uint32_t radix, bool isIntegerMode, wstring_view maxNumStr, uint32_t wordBitWidth, int maxDigits)
{
    // Convert from an integer into a character
    // This includes both normal digits and alpha 'digits' for radixes > 10
    auto chDigit = static_cast<wchar_t>((value < 10) ? (L'0' + value) : (L'A' + value - 10));

    CalcNumSec* pNumSec;
    size_t maxCount;
    if (m_hasExponent)
    {
        pNumSec = &m_exponent;
        maxCount = C_EXP_MAX_DIGITS;
    }
    else
    {
        pNumSec = &m_base;
        maxCount = maxDigits;
        // Don't include the decimal point in the count. In that way you can enter the maximum allowed precision.
        // Precision doesn't include decimal point.
        if (HasDecimalPt())
        {
            maxCount++;
        }
        // First leading 0 is not counted in input restriction as the output can be of that form
        // See NumberToString algorithm. REVIEW: We don't have such input restriction mimicking based on output of NumberToString for exponent
        // NumberToString can give 10 digit exponent, but we still restrict the exponent here to be only 4 digits.
        if (!pNumSec->IsEmpty() && pNumSec->value.front() == L'0')
        {
            maxCount++;
        }
    }

    // Ignore leading zeros
    if (pNumSec->IsEmpty() && (value == 0))
    {
        return true;
    }

    if (pNumSec->value.size() < maxCount)
    {
        pNumSec->value += chDigit;
        return true;
    }

    // if we are in integer mode, within the base, and we're on the last digit then
    // there are special cases where we can actually add one more digit.
    if (isIntegerMode && pNumSec->value.size() == maxCount && !m_hasExponent)
    {
        bool allowExtraDigit = false;

        if (radix == 8)
        {
            switch (wordBitWidth % 3)
            {
            case 1:
                // in 16 or 64bit word size, if the first digit is a 1 we can enter 6 (16bit) or 22 (64bit) digits
                allowExtraDigit = (pNumSec->value.front() == L'1');
                break;

            case 2:
                // in 8 or 32bit word size, if the first digit is a 3 or less we can enter 3 (8bit) or 11 (32bit) digits
                allowExtraDigit = (pNumSec->value.front() <= L'3');
                break;
            }
        }
        else if (radix == 10)
        {
            // If value length is at least the max, we know we can't add another digit.
            if (pNumSec->value.size() < maxNumStr.size())
            {
                // Compare value to substring of maxNumStr of value.size() length.
                // If cmpResult > 0:
                // eg. max is "127", and the current number is "20". first digit itself says we are out.
                // Additional digit is not possible

                // If cmpResult < 0:
                // Success case. eg. max is "127", and current number is say "11". The second digit '1' being <
                // corresponding digit '2', means all digits are possible to append, like 119 will still be < 127

                // If cmpResult == 0:
                // Undecided still. The case when max is "127", and current number is "12". Look for the new number being 7 or less to allow
                auto cmpResult = pNumSec->value.compare(0, wstring::npos, maxNumStr, 0, pNumSec->value.size());
                if (cmpResult < 0)
                {
                    allowExtraDigit = true;
                }
                else if (cmpResult == 0)
                {
                    auto lastChar = maxNumStr[pNumSec->value.size()];
                    if (chDigit <= lastChar)
                    {
                        allowExtraDigit = true;
                    }
                    else if (pNumSec->IsNegative() && chDigit <= lastChar + 1)
                    {
                        // Negative value case, eg. max is "127", and current number is "-12". Then 8 is also valid, as the range
                        // is always from -(max+1)...max in signed mode
                        allowExtraDigit = true;
                    }
                }
            }
        }

        if (allowExtraDigit)
        {
            pNumSec->value += chDigit;
            return true;
        }
    }

    return false;
}

bool CalcInput::TryAddDecimalPt()
{
    // Already have a decimal pt or we're in the exponent
    if (m_hasDecimal || m_hasExponent)
    {
        return false;
    }

    if (m_base.IsEmpty())
    {
        m_base.value += L'0'; // Add a leading zero
    }

    m_decPtIndex = m_base.value.size();
    m_base.value += m_decSymbol;
    m_hasDecimal = true;

    return true;
}

bool CalcInput::HasDecimalPt()
{
    return m_hasDecimal;
}

bool CalcInput::TryBeginExponent()
{
    // For compatibility, add a trailing dec point to base num if it doesn't have one
    TryAddDecimalPt();

    if (m_hasExponent) // Already entering exponent
    {
        return false;
    }

    m_hasExponent = true; // Entering exponent
    return true;
}

void CalcInput::Backspace()
{
    if (m_hasExponent)
    {
        if (!m_exponent.IsEmpty())
        {
            m_exponent.value.pop_back();

            if (m_exponent.IsEmpty())
            {
                m_exponent.Clear();
            }
        }
        else
        {
            m_hasExponent = false;
        }
    }
    else
    {
        if (!m_base.IsEmpty())
        {
            m_base.value.pop_back();
            if (m_base.value == L"0")
            {
                m_base.value.pop_back();
            }
        }

        if (m_base.value.size() <= m_decPtIndex)
        {
            // Backed up over decimal point
            m_hasDecimal = false;
            m_decPtIndex = 0;
        }

        if (m_base.IsEmpty())
        {
            m_base.Clear();
        }
    }
}

void CalcInput::SetDecimalSymbol(wchar_t decSymbol)
{
    if (m_decSymbol != decSymbol)
    {
        m_decSymbol = decSymbol;

        if (m_hasDecimal)
        {
            // Change to new decimal pt
            m_base.value[m_decPtIndex] = m_decSymbol;
        }
    }
}

bool CalcInput::IsEmpty()
{
    return m_base.IsEmpty() && !m_hasExponent && m_exponent.IsEmpty() && !m_hasDecimal;
}

wstring CalcInput::ToString(uint32_t radix)
{
    // In theory both the base and exponent could be C_NUM_MAX_DIGITS long.
    if ((m_base.value.size() > MAX_STRLEN) || (m_hasExponent && m_exponent.value.size() > MAX_STRLEN))
    {
        return wstring();
    }

    wstring result;

    if (m_base.IsNegative())
    {
        result = L'-';
    }

    if (m_base.IsEmpty())
    {
        result += L'0';
    }
    else
    {
        result += m_base.value;
    }

    if (m_hasExponent)
    {
        // Add a decimal point if it is not already there
        if (!m_hasDecimal)
        {
            result += m_decSymbol;
        }

        result += ((radix == 10) ? L'e' : L'^');
        result += (m_exponent.IsNegative() ? L'-' : L'+');

        if (m_exponent.IsEmpty())
        {
            result += L'0';
        }
        else
        {
            result += m_exponent.value;
        }
    }

    // Base and Exp can each be up to C_NUM_MAX_DIGITS in length, plus 4 characters for sign, dec, exp, and expSign.
    if (result.size() > C_NUM_MAX_DIGITS * 2 + 4)
    {
        return wstring();
    }

    return result;
}

Rational CalcInput::ToRational(uint32_t radix, int32_t precision)
{
    PRAT rat = StringToRat(m_base.IsNegative(), m_base.value, m_exponent.IsNegative(), m_exponent.value, radix, precision);
    if (rat == nullptr)
    {
        return 0;
    }

    Rational result{ rat };
    destroyrat(rat);

    return result;
}
